ASB physical addressing logic

yvesgoeleven opened the pull request:
- [x] explanation of the physical addressing logic layer in ASB transport,
- [x] it's relationship to naming conventions in V6 
- [x] incorporate changes introduced by #1470  (Sanitization)
- [x] samples on how to replace parts

yvesgoeleven commented:
Please review @Particular/tf-azure-service-bus-doco @Particular/docs-maintainers 

It's still a WIP because I know there were changes to the items under documentation which still need to be incorporated, can you review what needs to change @SeanFeldman 

Furthermore I somehow want to include samples on the parts to be replaced (ala replication/sharding strategy), but wonder if these should be snippets in the document or full blow samples, thoughts?

SimonCropp commented:
@yvesgoeleven i tried to clear up the guidance on samples https://github.com/Particular/docs.particular.net#samples

does that clear things up?

> Furthermore I somehow want to include samples on the parts to be replaced (ala replication/sharding strategy), but wonder if these should be snippets in the document or full blow samples, thoughts?

I think it depends on how complex it is to setup and how many moving pieces there are. if we only add example implementations to snippets, without a sample, how much effort would it be for someone to get started? would a downloadable solution greatly simplify things?

SimonCropp commented:
BTW even if we go with samples i think we should review + merge this PR and then do the samples as a diff PR.

SeanFeldman commented:
Please defer the review of this document until PR #1456 for sanitization is done. It will have a significant change on this doco.

SimonCropp commented:
@SeanFeldman will u be rebasing this?

SeanFeldman commented:
Would need to, yes @SimonCropp

weralabaj commented on nservicebus/azure-service-bus/addressing-logic.md:
```diff
@@ -0,0 +1,171 @@
+Prior to version 7 there was no explicit layer for handling this kind of logic, most assumptions such as assumptions on length limitations, legal characters, differences between paths and names etc... were sprinkled across the transport.
+
+Over time, subtle variations started to show in these assumptions as changes took place inside the implementation of certain namespace types. For example `Mixed` namespaces allowed paths and names up 290 characters, while `Messaging` namespaces are capped at 260. Next to that, customers also figured out ways to game the system (for valid reasons), like embedding `'/'` characters in endpoint names ...
+
```
> Prior to version 7 there was no explicit layer for handling this kind of logic, most assumptions such as assumptions on length limitations, legal characters, differences between paths and names etc... were sprinkled across the transport.

> Over time, subtle variations started to show in these assumptions as changes took place inside the implementation of certain namespace types. For example `Mixed` namespaces allowed paths and names up 290 characters, while `Messaging` namespaces are capped at 260. Next to that, customers also figured out ways to game the system (for valid reasons), like embedding `'/'` characters in endpoint names to create subfolders in the Azure Service Bus namespace registration system.

In Versions 6 and below, the *Physical Addressing Logic* implementation made a number of implicit assumptions regarding names, such as length limitations, legal characters, differences between paths and names, etc. There was no way to explicitly control those assumptions, and over time subtle variations started to show up.

For example `Mixed` namespaces allowed paths and names to have up to 290 characters, while `Messaging` namespaces are capped at 260 characters. Moreover, customers were sometimes forced to game the system for valid reasons. For example they needed to figure out how to embed slashes (`'/'`) in endpoint names to create subfolders in the Azure Service Bus namespace registration system.

weralabaj commented on nservicebus/azure-service-bus/addressing-logic.md:
```diff
@@ -0,0 +1,171 @@
+
+### Version 7
+
+In version 7 the configuration API allows to modify all behaviours and assumptions related to the addressing logic. This article will mainly focus on describing the layers inside the addressing logic and will show how to replace parts if further changes should occur, for a full list of out of the box options refer to [Full Configuration API](/nservicebus/azure-service-bus/configuration/configurat ...
```
> In version 7

In Versions 7 and above?

weralabaj replied:
Is the word "Layers" used in ASB in that context? It seems a bit surprising to me, because when I saw layers I thought about something like different levels of abstractions, e.g. physical vs logical address.

weralabaj commented on nservicebus/azure-service-bus/addressing-logic.md:
```diff
@@ -0,0 +1,171 @@
+
+In version 7 the configuration API allows to modify all behaviours and assumptions related to the addressing logic. This article will mainly focus on describing the layers inside the addressing logic and will show how to replace parts if further changes should occur, for a full list of out of the box options refer to [Full Configuration API](/nservicebus/azure-service-bus/configuration/configurat ...
+
+### Addressing layers
```
Is the word "Layers" used in ASB in that context? It seems a bit surprising to me, because when I saw layers I thought about something like different levels of abstractions, e.g. physical vs logical address.

SeanFeldman replied:
Perhaps "layer" is a wrong word to choose. 
Aspects?

yvesgoeleven replied:
Yeah, I lacked a good word here, maybe aspects is better if layer is too overloaded (aspects is overloaded as well though, but less)

SeanFeldman commented on nservicebus/azure-service-bus/addressing-logic.md:
```diff
@@ -0,0 +1,171 @@
+
+The sanitization layer is represented by an implementation of `ISanitizationStrategy`. 
+
+Out of the box there are 3 sanitization strategies
```
2 strategies. `AdjustmentSanitization` is converted into a sample (http://docs.particular.net/samples/azure/custom-sanitization/).

SeanFeldman commented on nservicebus/azure-service-bus/addressing-logic.md:
```diff
@@ -0,0 +1,171 @@
+The sanitization layer is represented by an implementation of `ISanitizationStrategy`. 
+
+Out of the box there are 3 sanitization strategies
+ * `AdjustmentSanitizationV6`: removes invalid characters according to `EntityNameValidationV6Rules`, uses MD5 hashing to reduce the length of an entity name if the maximum length is exceeded.
```
Renamed to `EndpointOrientedTopologySanitization`

SeanFeldman commented on nservicebus/azure-service-bus/addressing-logic.md:
```diff
@@ -0,0 +1,171 @@
+
+Out of the box there are 3 sanitization strategies
+ * `AdjustmentSanitizationV6`: removes invalid characters according to `EntityNameValidationV6Rules`, uses MD5 hashing to reduce the length of an entity name if the maximum length is exceeded.
+ * `AdjustmentSanitization` (default): removes invalid characters according to `EntityNameValidationRules`, uses SHA1 hashing to reduce the length of an entity name if the maximum length is exceeded.
```
Gone

yvesgoeleven commented:
@SimonCropp agreed, I referenced existing sample already, others can be linked in later

yvesgoeleven commented:
@Particular/docs-maintainers I incorporated the changes suggested during reviews and removed the WIP. If someone would like to do the final review and accept the PR, thanks @Particular/azure-service-bus-maintainers

SeanFeldman commented on nservicebus/azure-service-bus/addressing-logic.md:
```diff
@@ -0,0 +1,174 @@
+ * Sanitization: Determines how invalid entity names are cleaned up.
+ * Individualization: Determines how entity names are modified when used in different consumption modes.
+ * NamespacePartitioning: Determines how entities are partitioned across namespaces.
+ * Composition: Determines how entities are hierarchically composed inside a single namespace.
```
Composition is dealing with 2 types of entities that have a path (queues and topics). Other entities (subscriptions and rules) don't have paths and composition is not relevant. Perhaps the description should be focused on "path composition"?

SeanFeldman replied:
Also, for a reader who's not familiar with the concept, it's hard to understand what composition is all about.

weralabaj replied:
I'm +1 for something more like "path" or "namespace" composition.

SeanFeldman commented on nservicebus/azure-service-bus/addressing-logic.md:
```diff
@@ -0,0 +1,174 @@
+
+Out of the box there are 2 composition strategies
+ * `FlatComposition`: The entity is in the root of the namespace.
+ * `HierarchyComposition`: The entity is in a namespace hierarchy, at the location generated by the path generator.
```
Currently this is broken (https://github.com/Particular/NServiceBus.AzureServiceBus/issues/190) as there's no way to supply path generator for the strategy.
PR https://github.com/Particular/NServiceBus.AzureServiceBus/pull/225 will address it and add a configuration API to supply path generator.

SeanFeldman commented on nservicebus/azure-service-bus/addressing-logic.md:
```diff
@@ -0,0 +1,174 @@
+The validation aspect is represented by an implementation of `IValidationStrategy`. 
+
+Out of the box there are 2 validation strategies
+ * `EntityNameValidationV6Rules`: allows letters, numbers, periods (.), hyphens (-), and underscores (-)
```
To be consistent with other docs, the characters should be marked as code. F.e.
> periods (\`.\`)

this will render as: periods (`.`)

SeanFeldman commented on nservicebus/azure-service-bus/addressing-logic.md:
```diff
@@ -0,0 +1,174 @@
+ * `FlatComposition`: The entity is in the root of the namespace.
+ * `HierarchyComposition`: The entity is in a namespace hierarchy, at the location generated by the path generator.
+
+The default implementation of this strategy can be replaced by using the configuration API:
```
It is not clear which strategy is the default (`FlatComposition`)

SeanFeldman commented:
@yvesgoeleven added a few comments. One default is missing. 
A PR to address the issue with the `HierarchyComposition` is pending and once is merged, this document can be updated to state that for `HierarchyComposition` path generator can be registered using configuration API.

SimonCropp commented:
@yvesgoeleven @SeanFeldman FYI i pushed some minor cleanup things.

yvesgoeleven commented:
@SimonCropp thanks for addressing @SeanFeldman 's remark as well!

SeanFeldman commented:
PR for `HierarchyComposition` is merged. Path generator can be set using configuration API:

```c#
var compositionSettings = transport.Composition().UseStrategy<HierarchyComposition>();
compositionSettings.PathGenerator(Func<string,string>);
```

SimonCropp commented:
@yvesgoeleven @SeanFeldman is this one good to be merged? if not do u need anything from me ?

SeanFeldman commented:
@SimonCropp I'll need to update PR to be up-to-date based on code changes (https://github.com/Particular/docs.particular.net/pull/1456#issuecomment-218310642)

SeanFeldman commented:
@SimonCropp good to merge. I didn't have to change anything in the end.

SimonCropp commented on nservicebus/azure-service-bus/naming-conventions.md:
```diff
@@ -0,0 +1,28 @@
+---
+title: Azure Service Bus Transport Naming Conventions
```
@SeanFeldman want this in the menu?

SeanFeldman replied:
No, this one is only referenced from the main doc.

SimonCropp commented on nservicebus/azure-service-bus/addressing-logic.md:
```diff
@@ -0,0 +1,200 @@
+---
```
@SeanFeldman want this in the menu?

SeanFeldman replied:
Yes please. In the deserts section :smiley:

SeanFeldman commented:
@SimonCropp ready. Menu entry is added (after Crème brûlée)
